python-package/lightgbm/dask.py

```diff
@@ -338,6 +494,99 @@ def _train(
         for i in range(n_parts):
             parts[i]['init_score'] = init_score_parts[i]
 
+    # evals_set will to be re-constructed into smaller lists of (X, y) tuples, where
+    # X and y are each delayed sub-lists of original eval dask Collections.
+    if eval_set:
+
+        # find maximum number of parts in an individual eval set so that we can
+        # pad eval sets when they come in different sizes.
+        n_largest_eval_parts = max([x[0].npartitions for x in eval_set])
+
+        eval_sets = defaultdict(list)
+        if eval_sample_weight:
+            eval_sample_weights = defaultdict(list)
+        if eval_group:
+            eval_groups = defaultdict(list)
+        if eval_init_score:
+            eval_init_scores = defaultdict(list)
+
+        for i, (X_eval, y_eval) in enumerate(eval_set):
+            n_this_eval_parts = X_eval.npartitions
+
+            # when individual eval set is equivalent to training data, skip recomputing parts.
+            if X_eval is data and y_eval is label:
+                for parts_idx in range(n_parts):
+                    eval_sets[parts_idx].append(_DatasetNames.TRAINSET)
+            else:
+                eval_x_parts = _split_to_parts(data=X_eval, is_matrix=True)
+                eval_y_parts = _split_to_parts(data=y_eval, is_matrix=False)
+                for j in range(n_largest_eval_parts):
+                    parts_idx = j % n_parts
```

Comment:
Yeah sorry the `i` and `j` indexing is a lot to keep straight. You're probably are already aware that `i` is the iteration over the components of `eval_set` that are, at a minimum, to be comprised of the tuple `(X_i, y_i)`. And then `j` is the index of the iteration along the parts of one `(X_i, y_i)` along with possibly all of the extra distributed vectors like `eval_init_score[i]`.

I don't believe there is any overwriting going on - the logic is append-only. So when we go from `i=0` to `i=1` we get

```
i = 1

eval_sets = {
    0: [([x_e0_0, x_e0_3, x_e0_6], [y_e0_0, y_e0_3, y_e0_6]), ([x_e1_0], [y_e1_0])],
    1: [([x_e0_1, x_e0_4, None],         [y_e0_1, y_e0_4, None]),         ([None], [None])],
    2: [([x_e0_2, x_e0_5, None],         [y_e0_2, y_e0_5, None]),         ([None], [None])],
}
```

This is because of lines 530 and 531 that initiate a new tuple for the next component of `eval_set` (i.e. `i=1`):
```
if j < n_parts:
    eval_sets[parts_idx].append(([x_e], [y_e]))
```
So when `i` changes from 0 to 1 and then we re-enter the inner loop over `j`, then since `j=0, ...`, the `if j < n_parts:` condition is always reached first, and we append a new tuple, `([x_e1_0, y_e1_0])`. Then on line 531, once we've hit the round-robin criteria, we start appending `X_ij`'s to the first component of the last tuple contained in `eval_sets[parts_idx]` (append-only).

Reply:
AHHHHHHHH ok yes now I understand, thank you. And this is why the `None`-padding is important, because it guarantees that `len(eval_sets[i])` for any `i` will be equal to `len(eval_set)`. Got it!

Ok cool cool I'm good with this and can't think of a clearer way to do it. Could you just add a comment on the `if j < n_parts:` block briefly explaining this? Something like this

```python
if j < n_parts:
    # first time a chunk of this eval set is added to this part
    eval_sets[parts_idx].append(([x_e], [y_e]))
else:
    # appending additional chunks of this eval set to this part
    eval_sets[parts_idx][-1][0].append(x_e)
    eval_sets[parts_idx][-1][1].append(y_e)
```

Reply:
Yep for sure, sounds good